File: solutions.digamma.damas.jcr/src/main/java/solutions/digamma/damas/jcr/model/JcrManager.java
package solutions.digamma.damas.jcr.model;

import solutions.digamma.damas.common.AuthenticationException;
import solutions.digamma.damas.common.NotFoundException;
import solutions.digamma.damas.auth.Token;
import solutions.digamma.damas.jcr.session.SessionUser;
import solutions.digamma.damas.jcr.session.SessionBookkeeper;
import solutions.digamma.damas.jcr.session.SessionWrapper;
import solutions.digamma.damas.inspection.NotNull;

import javax.inject.Inject;
import java.util.logging.Logger;

/**
 * Generic manager abstract implementation.
 *
 * @author Ahmad Shahwan
 */
abstract public class JcrManager implements SessionUser {

    @Inject
    protected Logger logger;

    @Inject
    private SessionBookkeeper bookkeeper;

    @Override
    public @NotNull SessionWrapper getSession(@NotNull Token token)
            throws AuthenticationException {
        try {
            return this.bookkeeper.lookup(token);
        } catch (NotFoundException e) {
            throw new AuthenticationException("No session for token.", e);
        }
    }
}

File: solutions.digamma.damas.jcr/src/main/kotlin/solutions/digamma/damas/jcr/model/JcrManager.kt
package solutions.digamma.damas.jcr.model

import solutions.digamma.damas.common.AuthenticationException
import solutions.digamma.damas.common.NotFoundException
import solutions.digamma.damas.jcr.session.JcrTransaction
import solutions.digamma.damas.jcr.session.JcrSessionConsumer
import solutions.digamma.damas.logging.Logged
import java.util.logging.Logger
import javax.inject.Inject

/**
 * Generic manager abstract implementation.
 *
 * @author Ahmad Shahwan
 */
@Logged.Lifecycle
internal abstract class JcrManager : JcrSessionConsumer {

    @Inject
    protected lateinit var logger: Logger

    override val session
    @Logged
    @Throws(AuthenticationException::class)
    get() = try {
        JcrTransaction.get().getSession()
    } catch (e: NotFoundException) {
        throw AuthenticationException(e)
    }
}
